File: devstash/lib/models/Project.dart
class Project {
  String _title;
  String _desc;
  DateTime _createdAt;
  String _link;
  String _image;
  List<String> _technologies;
  List<String> _collaboraotrs;
  List<String> _hashTags;

  Project(this._title, this._desc, this._createdAt, this._link, this._image,
      this._technologies, this._collaboraotrs, this._hashTags);

  String get title => _title;
  set title(String value) => _title = value;

  String get desc => _desc;
  set desc(String value) => _desc = value;

  DateTime get createdAt => _createdAt;
  set createdAt(DateTime value) => _createdAt = value;

  String get link => _link;
  set link(String value) => _link = value;

  String get image => _image;
  set image(String value) => _image = value;

  List<String> get technologies => _technologies;
  set technologies(List<String> value) => _technologies = value;

  List<String> get collaboraotrs => _collaboraotrs;
  set collaboraotrs(List<String> value) => _collaboraotrs = value;

  List<String> get hashTags => _hashTags;
  set hashTags(List<String> value) => _hashTags = value;
}

File: backend/db/project.go
package db

import (
	"context"
	"fmt"

	"github.com/prakhar-5447/models"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func (store *MongoDBStore) Create_Project(ctx context.Context, project *Project) (*Project, error) {
	result, err := store.GetCollection("projects").InsertOne(ctx, project)
	if err != nil {
		return nil, err
	}

	project.ID = result.InsertedID.(primitive.ObjectID)

	return project, nil
}

func (store *MongoDBStore) Update_Project(ctx context.Context, projectID primitive.ObjectID, userID primitive.ObjectID, update models.ProjectRequest) (*Project, error) {
	collaboratorsID := make([]primitive.ObjectID, len(update.CollaboratorsID))
	for i, collabID := range update.CollaboratorsID {
		objectID, err := primitive.ObjectIDFromHex(collabID)
		if err != nil {
			return nil, err
		}
		collaboratorsID[i] = objectID
	}

	updateQuery := bson.M{
		"$set": bson.M{
			"image":           update.Image,
			"title":           update.Title,
			"url":             update.Url,
			"description":     update.Description,
			"technologies":    update.Technologies,
			"projectType":     update.ProjectType,
			"hashtags":        update.Hashtags,
			"collaboratorsID": collaboratorsID,
		},
	}

	result, err := store.projectsCollection.UpdateOne(ctx, bson.M{"_id": projectID, "userID": userID}, updateQuery)
	if err != nil {
		return nil, err
	}

	if result.ModifiedCount == 0 {
		return nil, fmt.Errorf("project not found or unauthorized access")
	}

	updatedProject := &Project{}
	err = store.projectsCollection.FindOne(ctx, bson.M{"_id": projectID}).Decode(updatedProject)
	if err != nil {
		return nil, err
	}

	return updatedProject, nil
}

func (store *MongoDBStore) Delete_Project(ctx context.Context, projectID primitive.ObjectID, userID primitive.ObjectID) (bool, error) {
	filter := bson.M{
		"userId": userID,
		"_id":    projectID,
	}

	result, err := store.projectsCollection.DeleteOne(ctx, filter)
	if err != nil {
		return false, err
	}

	if result.DeletedCount == 0 {
		return false, fmt.Errorf("project not found or unauthorized access")
	}

	return true, nil
}

func (store *MongoDBStore) Get_Projects_By_UserId(ctx context.Context, userID primitive.ObjectID) ([]*Project, error) {
	filter := bson.M{
		"userId": userID,
	}

	cursor, err := store.projectsCollection.Find(ctx, filter)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	projects := []*Project{}
	for cursor.Next(ctx) {
		project := &Project{}
		err := cursor.Decode(project)
		if err != nil {
			return nil, err
		}
		projects = append(projects, project)
	}

	return projects, nil
}

func (store *MongoDBStore) Get_Project_By_Id(ctx context.Context, projectID string) (*Project, error) {
	project := &Project{}
	objID, err := primitive.ObjectIDFromHex(projectID)
	if err != nil {
		return nil, err
	}

	err = store.projectsCollection.FindOne(ctx, bson.M{"_id": objID}).Decode(project)
	if err != nil {
		return nil, err
	}
	return project, nil
}
